gdal/alg/marching_squares/level_generator.h

```diff
@@ -0,0 +1,205 @@
+/******************************************************************************
+ *
+ * Project:  Marching square algorithm
+ * Purpose:  Core algorithm implementation for contour line generation.
+ * Author:   Oslandia <infos at oslandia dot com>
+ *
+ ******************************************************************************
+ * Copyright (c) 2018, Oslandia <infos at oslandia dot com>
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a
+ * copy of this software and associated documentation files (the "Software"),
+ * to deal in the Software without restriction, including without limitation
+ * the rights to use, copy, modify, merge, publish, distribute, sublicense,
+ * and/or sell copies of the Software, and to permit persons to whom the
+ * Software is furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included
+ * in all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS
+ * OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL
+ * THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+ * FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
+ * DEALINGS IN THE SOFTWARE.
+ ****************************************************************************/
+#ifndef MARCHING_SQUARE_LEVEL_GENERATOR_H
+#define MARCHING_SQUARE_LEVEL_GENERATOR_H
+
+#include <vector>
+#include <limits>
+#include <cmath>
+#include "utility.h"
+
+namespace marching_squares {
+
+template <class Iterator>
+class Range
+{
+public:
+    Range( Iterator b, Iterator e ) : begin_( b ), end_( e ) {}
+    Iterator begin() const { return begin_; }
+    Iterator end() const { return end_; }
+private:
+    Iterator begin_;
+    Iterator end_;
+};
+
+template <typename LevelIterator>
+class RangeIterator
+{
+public:
+    RangeIterator( const LevelIterator& parent, size_t idx ) : parent_( parent ), idx_( idx ) {}
+    std::pair<size_t, double> operator*() const
+    {
+        return std::make_pair( idx_, parent_.level( idx_ ) );
+    }
+    bool operator!=( const RangeIterator& other ) const
+    {
+        return idx_ != other.idx_;
+    }
+    const RangeIterator& operator++()
+    {
+        idx_++;
+        return *this;
+    }
+private:
+    const LevelIterator& parent_;
+    size_t idx_;
+};
+
+class FixedLevelIterator
```

Comment:
wouldn't FixedLevelRangeGenerator describe more the class (it is not an iterator itself, but can return an iterator) ?
Same for below IntervalLevelIterator and ExponentialLevelIterator